### PR TITLE
RFC: make `reshape` always share data

### LINKED ISSUE #4211: reshape should be consistent re copying or sharing data
Currently the documentation for `reshape` states:
```
julia> help(reshape)
Loading help data...
Base.reshape(A, dims)

   Create an array with the same data as the given array, but with
   different dimensions. An implementation for a particular type of
   array may choose whether the data is copied or shared.
```
Copying or sharing should be consistent across array types.  Perhaps use `reshape!` for a version that shares storage and save `reshape` for a version that copies the array's contents.

### PR DESCRIPTION
**Pull request by JeffBezanson**
Fixes #4211. I added `collect(itr, dims)` to make it easier to replace copying calls to reshape. I also had to revert #15251, since `squeeze` calls `reshape`. I assume we want it to have the same behavior. If not, I can add back a copying `squeeze` for sparse.

### CODE REVIEW

**Comment by timholy** [comment]
A lot of this is redundant with #15449. Why force people away from `reshape` and then encourage them back later?

File: base/array.jl [hunk 1]
```diff
@@ -208,16 +208,25 @@ _similar_for(c::AbstractArray, T, itr, ::HasShape) = similar(c, T, convert(Dims,
...
 
-Return an array of type `Array{element_type,1}` of all items in a collection.
+Return an array of type `Array{element_type,1}` of all items from an iterator.
 """
 collect{T}(::Type{T}, itr) = collect(Generator(T, itr))
 
+collect{T}(::Type{T}, itr::Dims) = collect(Generator(T, itr))
+
+"""
+    collect(iterator, dims)
```
Comment (tkelman) [inline base/array.jl]: needs an rst signature and a genstdlib run

**Comment by JeffBezanson** [comment]
@timholy That's a fair point. I would be fine with merging #15449 first, then using it to significantly cut down this change. I don't mind if there are a few performance corner cases.

**Comment by timholy** [comment]
Good.

IIRC (it's been a while), one of the few current sources of test failure in #15449 is that it goes whole-hog and even avoids calling the [C code that reshapes `Array`s](https://github.com/JuliaLang/julia/blob/2aa5d4d6f159f855a1ec6b2b113bef54153ca3ad/src/array.c#L163-L229). That code is not deleted in that PR, but if we go whole-hog it should be. One might view this as something we really want to do: I'll point out it lays the foundation for writing `Array` in julia-code and having `jltypes.c` simply define a `Buffer{T}` type. Alternatively, it could be viewed as an interesting demonstration that nothing "real" broke (either method-wise or performance-wise); point made, we could go back to using the C code on `Array` (at least for now).

The immediate decision we have to make is this: there are some tests that explicitly check the return type of `reshape([x for x in X], sz)`. Currently this returns an `Array`, but in #15449 it currently returns a `ReshapedArray{Array}`. There are at least three ways to fix this: (1) change the tests, (2) go back to relying on the C reshaping code for `Array`, or (3) define `[x for x in X]` to return an `Array` with the size of `X`. Which of these do you prefer?

**Comment by JeffBezanson** [comment]
I think we will do (3) in any case. I also think we should do (2) to make the transition smoother.

**Pull request closed without merging**